services/src/main/java/org/keycloak/services/scheduled/ClearExpiredAdminEvents.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.services.scheduled;
+
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.Time;
+import org.keycloak.events.EventStoreProvider;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.timer.ScheduledTask;
+
+public class ClearExpiredAdminEvents implements ScheduledTask {
```

Comment:
This does not belong to server-spi-private but to legacy-private module now.

Reply:
If it is not acceptable to have  clearExpiredAdminEvents() in the interface then there is no need for this scheduler in the generic scheduling. 
As some stores do not support expiry natively where would be the place to put this store specific scheduled job?

Reply:
Hi thanks for the hints. I did find the new location of the scheduled tasks before but was not sure if it would be fine to have an explicit dependency to the implementation of the store in this place.

Reply:
It is. I realized that you can also do the cleanup indirectly, say by having the `JpaEventStoreProvider` implement `InvalidationHandler` and sending an invalidation event from the scheduled task. Feel free to introduce a new `InvalidableObjectType` for that purpose in the legacy-private module, say `PeriodicEventInvalidation` could be the name. Given that the tasks are scheduled in the legacy-private module, this might even be the right option, since jpa classes are not available in legacy-private module.

See `InfinispanUserCacheProviderFactory` for an example.

Reply:
Thanks, this was the way I was thinking as well, I was about to find a place for a new TriggeredExpiration interface or something alike, but the InvalidationHandler seems fine, too. :)

Reply:
@hmlnarik will know better, but I would say it is ok to add `legacy-private` dependency to `model/jpa` module. Also, you should not check `InvalidationObjectType` in `JpaEventStoreProvider` but in `JpaEventStoreProviderFactory`. From the factory, you should then call a method on `JpaEventStoreProvider`.

Reply:
@mhajas Thank you for spotting this! Indeed, `JpaEventStoreProviderFactory` is the right one. @thpeter Apologies for the typo.

@thpeter `legacy-private` module is fine to be added to as a dependency to `model/jpa` module. Beware that the same is being added in #12686, so whichever goes into main first makes the second in the need for rebasing.

Reply:
#12686 is now merged